### DISCUSSION TITLE
Can't use docker registryURL to lookup dependencies

### DISCUSSION THREAD

**Discussion by Chrisman94** [renovatebot/renovate#16546] [Request Help]
Hi,

due to to the rate limiting of DockerHub we hosted a Docker-Cache as a registry(using a helm chart). I defined it as the registryUrl, but renovate has problems finding "node" for example. 

I tried to define it as
```
{
  "packageRules": [
    {
        "matchDatasources": ["docker"],
	"matchPackagePatterns": ["internalharbor"],
	"registryUrls": ["internalharborurl"]
    },
    {
      "matchDatasources": ["docker"],
      "excludePackagePatterns": ["internalharborurl"],
      "registryUrls": ["dockercacheurl"]
    }
  ]
}
```

In the lower part I defined the dockercache-url as registryUrl. When he tries to find the node:14-alpine image he looks for it in
**dockercacheurl/v2/node/tags/list?n=10000.**
The problem is the image can be found under 
**dockercacheurl/v2/library/node/tags/list?n=10000**

The last one should be the same path in the normal docker hub registry.
Any idea how we can fix this?

Thanks in advance

**Comment by PhilipAbed** [comment]
try adding "registryUrls":"dockercacheurl/library"
according to code:
```
const fullUrl = `${registryHost}/${dockerRepository}`;

urlToCall  = `${registryHost}/${dockerRepository}/tags/list?n=${limit}`;
```

**Reply by Chrisman94** [reply]
i tried that, but then he automaticaly adds a "v2".   "registryUrls":"dockercacheurl/v2/library/v2/node/tags..." I dont know why he adds that v2
\

**Reply by Chrisman94** [reply]
"v2/library' is part of our registryUrl

**Reply by PhilipAbed** [reply]
the call: ` url = ensurePathPrefix(url, '/v2');`

the function:
```
    export function ensurePathPrefix(url: string, prefix: string): string {
      const parsed = new URL(url);
      const fullPath = parsed.pathname + parsed.search;
      if (fullPath.startsWith(prefix)) {
        return url;
      }
      return parsed.origin + prefix + fullPath;
    }
```

**Reply by PhilipAbed** [reply]
try to unit test that :D with ur URL and see why it adds
or just send me your URL and i can do it for you

the url is : `${registryHost}/${dockerRepository}/tags/list?n=${limit}`;`
before it enters that method

**Reply by PhilipAbed** [reply]
Basically if the `new URL(url);` returns a `Pathname` that doesnt start with `/v2` it fails your url adding this extra prefix

that means when we are parsing your dockercacheurl/v2/library
There's something in between like dockercacheurl/something/v2/library

**Reply by PhilipAbed** [reply]
for some reason the Pathname you got is `library/v2/` and not `v2/library` check your logs/URL see why :/

**Reply by PhilipAbed** [reply]
![image](https://user-images.githubusercontent.com/42116482/178527512-20d7e104-3af1-4860-9be1-d90b824c545b.png)
here's an example, i set my `URL/library/v2` and i got `/v2/library/v2 like you `

**Reply by rarkins** [reply]
I think someone in the team is already working on this topic of adding in library/ to the url. @MaronHatoum ?

**Comment by Chrisman94** [comment]
Thanks again. I checked the code as well and if it doesnt found it under the normal path he should try it again with +/library. So it should work actually.

I found out that the problem happens way earlier in getAuthHeaders:
```
   if (
      apiCheckResponse.statusCode !== 401 ||
      !is.nonEmptyString(apiCheckResponse.headers['www-authenticate'])
    ) {
      logger.warn(
        { apiCheckUrl, res: apiCheckResponse },
        'Invalid registry response'
      );
      return null;
    }
```

He calls getAuthHeaders() with apicheckUrl =` https://dockercacheurl/v2/node/tags/list?n=10000`

He should call it with `https://dockercacheurl/v2` only, then it would work fine.

**Reply by PhilipAbed** [reply]
welcome anytime

**Comment by Chrisman94** [comment]
This code checks the full image path for AuthHeaders, but this throws a 404 in my case, because the library-prefix is missing. It should only check the registryHost in my opinion. Is this a issue or do i have to configure something?
So as i see it: The Problem lies in 

`private async getDockerApiTags(
    registryHost: string,
    dockerRepository: string
  ): Promise<string[] | null> {
    let tags: string[] = [];
    // AWS ECR limits the maximum number of results to 1000
    // See https://docs.aws.amazon.com/AmazonECR/latest/APIReference/API_DescribeRepositories.html#ECR-DescribeRepositories-request-maxResults
    const limit = ecrRegex.test(registryHost) ? 1000 : 10000;
    let url:
      | string
      | null = `${registryHost}/${dockerRepository}/tags/list?n=${limit}`;
    url = ensurePathPrefix(url, '/v2');
    const headers = await getAuthHeaders(
      this.http,
      registryHost,
      dockerRepository,
      url
    );
    ....`

**Reply by PhilipAbed** [reply]
you mean `getAuthHeaders` fails because it takes the host, and adds `/v2` to it only? no library? but it should be `/v2/library` then it would work?

**Reply by Chrisman94** [reply]
No the problem is that getDockerApiTags passes  `${registryHost}/${dockerRepository}/tags/list?n=${limit}` as url to ` getAuthHeaders`, which overrides apiCheckUrl

In my case he passes `https://dockercacheurl/v2/node/tags/list?n=10000` which gives a HTTP404 of course
It should pass nothing in my opinion, so it uses the default of `getAuthHeaders` which is `${registryHost}/v2/` , this gives a HTTP200.

Just like 
  ```
private async getManifestResponse(
    registryHost: string,
    dockerRepository: string,
    tag: string,
    mode: 'head' | 'get' = 'get'
  ): Promise<HttpResponse | null> {
    logger.debug(
      `getManifestResponse(${registryHost}, ${dockerRepository}, ${tag})`
    );
    try {
      const headers = await getAuthHeaders(
        this.http,
        registryHost,
        dockerRepository
      );

```

**Reply by PhilipAbed** [reply]
but the `getAuthHeaders` is only getting the `registryHost`, not the full URL

**Reply by Chrisman94** [reply]
No it gets the full url, because we pass 4 parameters to the function. so `url` overwrites the default of `apiCheckUrl` in `getAuthHeaders`

see

```
private async getDockerApiTags(
    registryHost: string,
    dockerRepository: string
  ): Promise<string[] | null> {
    let tags: string[] = [];
    // AWS ECR limits the maximum number of results to 1000
    // See https://docs.aws.amazon.com/AmazonECR/latest/APIReference/API_DescribeRepositories.html#ECR-DescribeRepositories-request-maxResults
    const limit = ecrRegex.test(registryHost) ? 1000 : 10000;
    let url:
      | string
      | null = `${registryHost}/${dockerRepository}/tags/list?n=${limit}`;
    url = ensurePathPrefix(url, '/v2');
    const headers = await getAuthHeaders(
      this.http,
      registryHost,
      dockerRepository,
      url
    );
```


getAuthHeaders

```
export async function getAuthHeaders(
  http: Http,
  registryHost: string,
  dockerRepository: string,
  apiCheckUrl = `${registryHost}/v2/`
): Promise<OutgoingHttpHeaders | null> {
```

**Reply by Chrisman94** [reply]
@rarkins this is the actual problem

**Reply by Chrisman94** [reply]
I dont think we should use the full url for the purpose of getting Auth Headers.

**Reply by viceice** [reply]
it must be used, otherwise other registries will fail, check the git history and you found why we do that

**Reply by Chrisman94** [reply]
Okay im fine with it if that is the case. But then we should retry it with +library.

something like that:

```
private async getDockerApiTags(
    registryHost: string,
    dockerRepository: string
  ): Promise<string[] | null> {
    let tags: string[] = [];
    // AWS ECR limits the maximum number of results to 1000
    // See https://docs.aws.amazon.com/AmazonECR/latest/APIReference/API_DescribeRepositories.html#ECR-DescribeRepositories-request-maxResults
    const limit = ecrRegex.test(registryHost) ? 1000 : 10000;
    let url:
      | string
      | null = `${registryHost}/${dockerRepository}/tags/list?n=${limit}`;
    url = ensurePathPrefix(url, '/v2');
    const headers = await getAuthHeaders(
      this.http,
      registryHost,
      dockerRepository,
      url
    );
    if(headers === null){
       url = `${registryHost}/library/${dockerRepository}/tags/list?n=${limit}`;
       url = ensurePathPrefix(url, '/v2');
       headers = await getAuthHeaders(
           this.http,
           registryHost,
           dockerRepository,
           url
    );

```

or is this the thing you are currently working at?

**Reply by Chrisman94** [reply]
If you can't fix it, is there a workaround for us?

**Reply by PhilipAbed** [reply]
check [this](https://github.com/renovatebot/renovate/pull/16509/files) pull request, 
if getAuthHeaders fail or gets 404, 
after this fix it will try to add Library to the url, does this solve your issue?

**Reply by PhilipAbed** [reply]
fallback exists already in 
 https://github.com/renovatebot/renovate/blob/2b582140c26d3c30b3792911e7ea20c21a8bcf7e/lib/modules/datasource/docker/index.ts#L766-L774
 
 we just didnt handle the error correctly, when the PR is merged i think your issue will be solved

**Reply by Chrisman94** [reply]
The pull request probably fixes my issue.
But if I see it right, when the page is not found he just returns an empty object.
Wouldn't it be wiser to retry it with +"library/" so we actually call a (probably) existing resource? There could be authentication needed. Or is it not possible to get a 404 when there is authentication needed?

**Reply by PhilipAbed** [reply]
no idea, according to the tests maroon did it works fine after that change

**Reply by PhilipAbed** [reply]
worst case scenario you open a new issue if it doesnt work :D
cross fingers

**Reply by Chrisman94** [reply]
@MaronHatoum yeah i understand the code, and yeah it will work fine ;) I was just curious if there would be always a 401 when the ressource is not existing but authentication would be needed to access resources in general. 
How long do you think it takes to merge the PR?

**Reply by MaronHatoum** [reply]
Hi @Chrisman94 
it was just created today as soon as passes code review it will be merged.

**Reply by Chrisman94** [reply]
Can I do the code-review? :D

**Reply by MaronHatoum** [reply]
sure, everyone is welcome to review and suggest improvements,
but the maintainer's approval is mandatory.

**Reply by Chrisman94** [reply]
Is It already available in the latest docker image of renovate?
Error still exists currently in my repository. 

I get a 404 in getAuthHeaders because the library prefix is missing. There is no retry visible in the logs.

**Reply by MaronHatoum** [reply]
hi @Chrisman94 
this release 32.127.1 contains the bug fix. 
please check the release you are using.
